hazelcast/src/test/java/com/hazelcast/test/AssertEnabledFilterRule.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.test;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * Prevents a test annotated with {@link RequireAssertEnabled} from running when Java assertions are disabled.
+ * Typically such a test will expect an {@code AssertionError} to be thrown.
+ */
+public class AssertEnabledFilterRule implements TestRule {
+    @Override public Statement apply(final Statement base, final Description description) {
+        if (description.getAnnotation(RequireAssertEnabled.class) == null) {
+            return base;
+        }
+        return new Statement() {
+            @Override public void evaluate() throws Throwable {
+                boolean assertEnabled = false;
```

Comment:
**This is enough.**

On the other hand, in fact assertions can be specified classloader based, package based or even class based.
So in theory this means that, assertion might be enabled in here but disabled at the tested class itself (`Long2LongElasticMapHsa` in our case at another PR and also I know this is very edge case that we do not need to consider here).

So the another way (but complex) might be like this:

``` java
@Retention(RUNTIME)
@interface RequireAssertEnabled {
    
    Class<?> targetClass() default Object.class;
    
}

class AssertEnabledFilterRule implements TestRule {
    
    @Override 
    public Statement apply(final Statement base, final Description description) {
        final RequireAssertEnabled assertEnabled = 
                description.getAnnotation(RequireAssertEnabled.class);
        if (assertEnabled == null) {
            return base;
        }
        return new Statement() {
            @Override 
            public void evaluate() throws Throwable {
                boolean assertionEnabled = false;
                Class<?> targetClass = assertEnabled.targetClass();
                if (Object.class.equals(targetClass)) {
                    // We are considering test class's assertion status
                    assertionEnabled = description.getTestClass().desiredAssertionStatus();
                } else {
                    // We are considering specified class's assertion status
                    assertionEnabled = targetClass.desiredAssertionStatus();
                }
                if (assertionEnabled) {
                    base.evaluate();
                } else {
                    System.err.println("WARNING! Test cannot run when Java assertions are not enabled (java -ea ...): "
                            + description.getDisplayName());
                }
            }
        };
    }

}
```